Backport H5Dchunk_iter to 1.10 branch

mkitti opened the pull request:
This backports H5Dchunk_iter #6 to the 1.10 branch from the 1.13 development branch.
xref: https://forum.hdfgroup.org/t/possibility-of-backporting-h5dchunk-iter-to-1-12/9971/2

To do:
- [x] Backport tests from `test/chunk_iter.c`
- [x] Backport #1969 to fix offsets that were not multiplied by the chunk dimensions (see #1419)
- [x] Finish backport of `H5Dchunk_iter` to 1.12: #1970

mkitti committed:
Backport H5Dchunk_iter to 1.10 branch

mkitti committed:
Add some accessory files, test still needs work

mkitti commented:
The test failure encountered in #666 may be an issue here as well

derobins commented:
This needs to go to 1.12 first. We move features down the chain of maintenance branches.

mkitti committed:
Apply proper formatting, and fix compile errors

mkitti committed:
Remove const from H5D__chunk_iter as per #1700

mkitti committed:
H5Dchunk_iter now passes offsets in units of dataset elements, fix #1419 (#1969)

* H5Dchunk_iter now passes chunk dimension scaled offsets, fix #1419

* Update docs for H5Dchunk_iter, H5Dget_chunk_info*

Modified description for `H5Dchunk_iter`, `H5Dget_chunk_info`, and `H5Dget_chunk_info_by_coord` to the following

 * offset          Logical position of the chunk’s first element in units of dataset elements
 * filter_mask  Bitmask indicating the filters used when the chunk was written
 * size             Chunk size in bytes, 0 if the chunk does not exist

mkitti committed:
Align arg types of H5D_chunk_iter_op_t with H5Dget_chunk_info (#2074)

* Align arg types of H5D_chunk_iter_op_t with H5Dget_chunk_info

* Modify chunk_info test to for unsigned / hsize_t types

* Fix types in test

mkitti committed:
Add test_basic_query, helper functions to test/chunk_info.c 1_10

mkitti committed:
Fix regression of #1419

mkitti committed:
Add a note about return fail in 1.12 and older for invalid chunk index

github-actions[bot] committed:
Committing clang-format changes

mkitti committed:
Run clang-format on test/chunk_info.c

mkitti committed:
Merge branch 'mkitti/h5dchunk_iter_1_10' of github.com:mkitti/hdf5 into mkitti/h5dchunk_iter_1_10

mkitti commented on test/chunk_info.c:
```diff
@@ -1666,6 +1806,280 @@ test_chunk_info_version2_btrees(char *filename, hid_t fapl)
+        ret = H5Dget_chunk_info(dset, dspace, chk_index, out_offset, &read_flt_msk, &addr, &size);
+    }
+    H5E_END_TRY;
+    /* In HDF5 1.10 this will not fail. It will fail in 1.12 and older. */
+    /*
+    if (ret != FAIL)
+        TEST_ERROR;
+    */
```
Trying to fix this creates new issues on HDF5 1.10.

mkitti commented:
I backported the `test_basic_query` in `test/chunkinfo.c`. This mostly involved changing the names of some constants to older variants. Also an actual dataspace identifier instead of `H5S_ALL` needed to be used.

I noticed some difference in return value behavior of `H5Dget_chunk_info` when trying to get information on a chunk greater than the number of chunks. Thus I had to disable one test that I noted above.

Since the HDF5 1.12 version of `H5Dchunk_iter` is advancing, the time for this may be appropriate now.
